### PR TITLE
feat: Migrate to controller-runtime

### LINKED ISSUE #40: Migrate jx controller to controller-runtime
In the same spirit as https://github.com/jenkins-x/lighthouse/issues/933

There’s a lot of boilerplate code in the jx controller.

How about migrating the controller to controller-runtime ? This would simplify the implementation enormously and would be a lot easier to maintain.

### LINKED ISSUE #45: Is config agent and watcher used in the controller ?
https://github.com/jenkins-x/lighthouse-jx-controller/blob/5ebf5c01c83d13433a81ef3c6211ac0a8251fdb7/pkg/engines/jx/controller.go#L90-L115

Is config agent and watcher used ?
They are instantiated but they don't seem to be used and i'm not sure if they are needed.

### PR DESCRIPTION
**Pull request by eddycharly**
This PR migrates the controller to controller-runtime.

Closes #40 
Closes #45

### CODE REVIEW

**Comment by abayer** [comment]
https://github.com/jenkins-x/jx-api/blob/master/go.mod#L9 is the problem, I believe.

**Comment by abayer** [comment]
Hmm, adding `github.com/go-logr/logr v0.2.1-0.20200730175230-ee2de8da5be6` to `go.mod` seems to fix things.

**Comment by eddycharly** [comment]
Wow that's it, well done !

Shall i pin the dependency to `v0.1.0` ?

**Comment by abayer** [comment]
Let's see...I'm fine either way, honestly.

**Comment by eddycharly** [comment]
Ok, it builds, let's see what CI says...

**Comment by abayer** [comment]
Yeah, it's compiling but not actually working. =) The `PipelineActivity` isn't getting owner information on it, which regrettably does not surprise me - it's actually getting created by some jx code we call here, so we'd have to make a change to it after creation to add the owner info.

**Comment by eddycharly** [comment]
Ah yes, but we never get access to the `PipelineActivity` (neither `mpClient.Create` or `mpClient.Apply` it).

I imagine `PipelineActivity` is a more abstract resource that the controller does not access and can only watch ?

**Comment by abayer** [comment]
Yeah, we could get it from the `activityKey.Name` and then update it, but it's probably better to just watch the `PipelineActivity`s.

**Comment by eddycharly** [comment]
It's watching `PipelineActivity`s, i will change the indexer rule tu use labels instead of owner ref.

**Comment by eddycharly** [comment]
I made the changes (trying to use pipeline activity name) but not sure if it works better.

Let me know if you see something interesting in the log

**Comment by abayer** [comment]
I’ll try to dig into this today and figure out why it’s not getting the activity...

**Comment by abayer** [comment]
@eddycharly At this point, it's probably worth trying to turn the unit tests back on and get them to pass.

**Comment by eddycharly** [comment]
Yeah, i there's one and it's passing. I will uncomment others too.
Did you see something in the logs ?

**Comment by abayer** [comment]
No, nothing useful - it just looks like it's not finding the `PipelineActivity` for the `LighthouseJob`, but I can't tell why not.

**Comment by abayer** [comment]
So I figured one thing out - you switched to `gopkg.in/yaml.v1`, which breaks the templating of the report URL. Switching back to `sigs.k8s.io/yaml` fixes that - `gopkg.in/yaml.v1` ends up resulting in the params map having lower-case keys. So that'll fix `update-job` test case, and the `no-job-for-activity` test case doesn't actually apply any more. I'm pushing this change to your branch to see if this turns out to be all we needed...

**Comment by eddycharly** [comment]
🤞

**Comment by abayer** [comment]
Nope, didn't help. The watch doesn't seem to be working, I guess.

**Comment by abayer** [comment]
We could change things so that after the `r.mpClient.Apply(...)`, we `Get` the `PipelineActivity` with the name `activityKey.Name` and then update it with the owner reference...but I'm not sure that would actually be desirable, since it's likely that the `PipelineActivity` is expected to be kept around longer than the `LighthouseJob`. So I guess we gotta figure out the watch...

**Comment by eddycharly** [comment]
Is the `PipelineActivity` created ?
That's difficult to reproduce because i don't have a jx install at hand, it's a guessing game.

If the `PipelineActivity` is created, we know it's an issue with the watch....

**Comment by abayer** [comment]
Yes, the `PipelineActivity` does get created.

**Comment by eddycharly** [comment]
ok, i will add a log in the watch handler... maybe the handler is not finding the job

**Comment by eddycharly** [comment]
I added a log, now we should know if watching works but fails to find the underlying job or don't watch at all

File: pkg/engines/jx/controller.go [hunk 1]
```diff
@@ -1,570 +1,275 @@
...
 
 const (
 	controllerName           = "jx-controller"
-	activityKeyPrefix        = "activity"
-	jobKeyPrefix             = "job"
 	defaultTargetURLTemplate = "{{ .BaseURL }}/teams/{{ .Team }}/projects/{{ .Owner }}/{{ .Repository }}/{{ .Branch }}/{{ .Build }}"
 	baseTargetURLEnvVar      = "LIGHTHOUSE_REPORT_URL_BASE"
 	targetURLTeamEnvVar      = "LIGHTHOUSE_REPORT_URL_TEAM"
-	targetURLTemplateEnvVar  = "LIGHTHOUSE_REPORT_URL_TEMPLATE"
+	pipelineActivityKey      = ".metadata.pipelineActivity"
```
Comment (abayer) [inline pkg/engines/jx/controller.go:35]: So I'm confused as to what exactly this is - is it the name of a real field, or is it a field we're creating?

Comment (eddycharly) [reply pkg/engines/jx/controller.go:35]: It’s the key used by the indexer to lookup resources from the cache

Comment (abayer) [reply pkg/engines/jx/controller.go:35]: Ok, so it's just an arbitrary string?

Comment (eddycharly) [reply pkg/engines/jx/controller.go:35]: Yes, preferably not something that already exists.

Comment (abayer) [reply pkg/engines/jx/controller.go:35]: Ok, I was quite confused. =)

**Comment by eddycharly** [comment]
Doh, I think I got it !

**Comment by abayer** [comment]
I'm seeing the `PipelineActivity` and `LighthouseJob` in the log now, so I guess the watch is working but the list isn't maybe?

File: pkg/engines/jx/controller.go [hunk 2]
```diff
@@ -141,7 +141,7 @@ func (r *LighthouseJobReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 
 	// get job's pipeline activities
 	var pipelineActivityList jxv1.PipelineActivityList
-	if err := r.client.List(ctx, &pipelineActivityList, client.InNamespace(req.Namespace), client.MatchingFields{pipelineActivityKey: req.Name}); err != nil {
+	if err := r.client.List(ctx, &pipelineActivityList, client.InNamespace(req.Namespace), client.MatchingFields{pipelineActivityKey: job.Status.ActivityName}); err != nil {
```
Comment (abayer) [inline pkg/engines/jx/controller.go:144]: ah-ha!

**Comment by eddycharly** [comment]
😓

**Comment by abayer** [comment]
`ghe` failure was a burp in the cluster's athens proxy. Everything else looks...very good, actually. =)

**Comment by eddycharly** [comment]
Finally !!
I thought it would be easier, it looked so much like the tekton controller...

I need to double check my changes and adapt the last unit tests, after that it should be good.

**Comment by abayer** [comment]
@eddycharly This has gotten out of date thanks to the config/job/etc refactoring. =)

**Comment by eddycharly** [comment]
Gonna look at it.

**Comment by eddycharly** [comment]
Fixed

**Pull request merged**